evalml/automl/engines/engine_base.py

```diff
@@ -0,0 +1,212 @@
+import sys
+import time
+import traceback
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+
+import numpy as np
+import pandas as pd
+from sklearn.model_selection import train_test_split
+
+from evalml.exceptions import PipelineScoreError
+from evalml.model_family import ModelFamily
+from evalml.pipelines import BinaryClassificationPipeline
+from evalml.problem_types import ProblemTypes
+from evalml.utils.gen_utils import (
+    _convert_to_woodwork_structure,
+    _convert_woodwork_types_wrapper
+)
+from evalml.utils.logger import get_logger, update_pipeline
+
+logger = get_logger(__file__)
+
+
+class EngineBase(ABC):
+    """ Base class for engines, which handles the fitting and evaluation of pipelines."""
+    def __init__(self):
+        """This class represents an "engine" for AutoML, which handles the evaluation of a list of pipelines generated from an AutoML search.
+
+        To use this interface, you must define an `evaluate_batch` method and an `evaluate_pipeline` method.
+        """
+        self.name = "Base Engine"
+        self.X = None
+        self.y = None
+        self.automl = None
+
+    def load_data(self, X, y):
+        """Loads the data to fit the pipeline on. Required to run `_compute_cv_scores`."""
+        self.X = X
+        self.y = y
+
+    def load_search(self, search_obj):
+        """Loads the current AutoML state. This includes information such as the search parameters and objectives, which is required to run `_compute_cv_scores`."""
+        self.automl = search_obj
+
+    @abstractmethod
```

Comment:
How come we need both `evaluate_batch` and `evaluate_pipeline`? I think it would be nice if there was only one `evaluate_batch` that lived in the base class that called `evaluate_pipeline` on each pipeline in the batch. I think having only one `evaluate_batch` implementation will help with maintainability in the long run:
- It'll help us keep the `KeyboardInterrupt` in only one place as opposed to every engine implementation. 
- It will also help us keep the `if self.X is None` checks in one place as well

I think the trade-off is that this means that `evaluate_pipeline` should always return a "future"-like object. We would have to create an `EngineResult` base class to create a single api between the results of all the engines. 

So it's not without a cost but it could be worth it to reduce the code duplication between the `evaluate_batch` and `evaluate` of different engines.

Reply:
I agree with the two points you made for just keeping `evaluate_batch`. My original reasoning for having a separate `evaluate_batch` and `evaluate_pipeline` function is that there would be separate overhead with processing a batch function that isn't needed in processing a single pipeline. However, in practice, there really isn't that much of a difference especially since for Dask we are distributing the dataset across workers in `load_data`. It also allows for a separate method for processing for the baseline search pipeline and `add_to_rankings`, although this is moot since we could just specify a "batch" with one pipeline. I'll take a pass at refactoring so that we just have `evaluate_batch`.

I also agree on the `EngineResult` base class. We're currently returning three things, with one of them being a big dictionary with the numerical results. It might also make sending back error messages (such as in the case of search termination) more explicit to automl search.